clash-lib/src/Clash/Primitives/Util.hs

```diff
@@ -123,29 +123,29 @@ addGuards
   -> ResolvedPrimMap
 addGuards = foldl go
  where
-  lookupPrim :: TS.Text -> ResolvedPrimMap -> Maybe ResolvedPrimitive
-  lookupPrim nm primMap = join (extractPrim <$> HashMapStrict.lookup nm primMap)
+  lookupPrim
+    :: TS.Text
+    -> ResolvedPrimMap
+    -> Maybe ([PrimitiveWarning], ResolvedPrimitive)
+  lookupPrim nm primMap = do
+    guardedPrim <- HashMapStrict.lookup nm primMap
+    prim <- extractPrim guardedPrim
+    return (extractWarnings guardedPrim, prim)
 
   go primMap (nm, guard) =
     HashMapStrict.insert
       nm
       (case (lookupPrim nm primMap, guard) of
-        (Nothing, HasBlackBox _) ->
+        (Nothing, DontTranslate) -> DontTranslate
+        (Nothing, HasBlackBox _ ()) ->
           error $ "No BlackBox definition for '" ++ TS.unpack nm ++ "' even"
                ++ " though this value was annotated with 'HasBlackBox'."
-        (Nothing, WarnNonSynthesizable _ _) ->
-          error $ "No BlackBox definition for '" ++ TS.unpack nm ++ "' even"
-               ++ " though this value was annotated with 'WarnNonSynthesizable'"
-               ++ ", implying it has a BlackBox."
-        (Nothing, WarnAlways _ _) ->
-          error $ "No BlackBox definition for '" ++ TS.unpack nm ++ "' even"
-               ++ " though this value was annotated with 'WarnAlways'"
-               ++ ", implying it has a BlackBox."
         (Just _, DontTranslate) ->
           error (TS.unpack nm ++ " was annotated with DontTranslate, but a "
-                                 ++ "BlackBox definition was found anyway.")
-        (Nothing, DontTranslate) -> DontTranslate
-        (Just p, g) -> fmap (const p) g)
+                              ++ "BlackBox definition was found anyway.")
+        (Just (ws1, p), HasBlackBox ws2 ()) ->
+          HasBlackBox (ws1 ++ ws2) p
```

Comment:
It's been a while since I've looked at this code, but I'm a little bit surprised by this. Do you know why these warnings are merged this late? Shouldn't they have been merged in all the compilation / discover steps we do?

Reply:
I'm not sure what you refer to as the compilation / discover steps, or what a better place for merging the warnings could be, but the reason that they need to be merged this late is that [just a couple of lines up](https://github.com/clash-lang/clash-compiler/blob/1bde4831c94e38a4359be2c9c5fbab8e871569f6/clash-lib/src/Clash/Primitives/Util.hs#L103-L104) a warning field of a primitive template yields a new `PrimitiveGuard` with a warning, and there may already be a `PrimitiveGuard` annotation with warnings.

Reply:
Aha, I misremember how this worked. Looks perfect!